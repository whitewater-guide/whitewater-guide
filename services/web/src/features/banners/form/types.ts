import { BannerInput, Overwrite } from '@whitewater-guide/commons';
import { LocalPhoto } from '../../../utils/files';

export type BannerFormData = Overwrite<
  BannerInput,
  {
    extras: string | null;
    source: string | LocalPhoto;
  }
>;

export interface RouterParams {
  bannerId?: string;
}
