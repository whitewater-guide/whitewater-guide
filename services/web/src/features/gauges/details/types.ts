import { WithGauge } from '@whitewater-guide/clients';

export type GaugeDetailsProps = WithGauge;
