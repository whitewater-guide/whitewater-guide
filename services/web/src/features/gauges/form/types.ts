import { Coordinate3d, GaugeInput, Overwrite } from '@whitewater-guide/commons';

export interface RouterParams {
  sourceId: string;
  gaugeId?: string;
}

export interface GaugeLocationForm {
  id?: string | null;
  name?: string | null;
  description?: string | null;
  coordinates: Coordinate3d | null;
  kind?: string;
}

export type GaugeFormData = Overwrite<
  GaugeInput<string | null>,
  { location: GaugeLocationForm | null }
>;
