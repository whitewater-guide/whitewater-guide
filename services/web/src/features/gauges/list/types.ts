import { WithGaugesList } from '@whitewater-guide/clients';
import { Source } from '@whitewater-guide/commons';
import { RouteComponentProps } from 'react-router-dom';
import { WithDeleteMutation } from '../../../apollo';
import { WithToggleGauge } from './withToggleGauge';

export interface GaugesListOuterProps {
  source: Source;
}

export type GaugesListInnerProps = GaugesListOuterProps &
  WithGaugesList &
  WithDeleteMutation<'removeGauge'> &
  WithToggleGauge &
  RouteComponentProps<{ sourceId: string }>;
