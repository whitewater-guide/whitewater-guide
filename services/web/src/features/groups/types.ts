import { WithGroupsList } from '@whitewater-guide/clients';
import { GroupInput } from '@whitewater-guide/commons';

export interface WithGroupMutations {
  upsertGroup: (group: GroupInput) => void;
  removeGroup: (id: string) => void;
}

export type GroupsFormProps = WithGroupsList & WithGroupMutations;
