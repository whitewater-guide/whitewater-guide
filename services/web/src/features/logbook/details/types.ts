export interface RouterParams {
  descentId?: string;
}
