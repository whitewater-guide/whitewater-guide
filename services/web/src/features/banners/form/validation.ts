import { getLocalPhotoSchema } from '@whitewater-guide/clients';
import {
  BannerInputSchema,
  BannerResolutions,
  yupTypes,
} from '@whitewater-guide/commons';
import * as yup from 'yup';
import { BannerFormData } from './types';

const ImageSourceSchema = yup.object().test({
  name: 'is-correct-image-banner-source',
  message: 'web:banners.invalidSource',
  test(v: any) {
    const placement = (this.options.context as any).placement;
    const resolution = BannerResolutions.get(placement);
    const localPhotoSchema = getLocalPhotoSchema({
      mpxOrResolution: resolution,
    });
    try {
      localPhotoSchema.validateSync(v);
      return true;
    } catch (e) {
      return this.createError({
        message: (e as yup.ValidationError).message,
      });
    }
  },
});

const WebViewSourceSchema = yup
  .string()
  .url()
  .required();

export const BannerFormSchema: yup.Schema<
  BannerFormData
> = BannerInputSchema.clone().shape({
  extras: yupTypes.jsonString().nullable(),
  regions: yup.array().of(yupTypes.namedNode()),
  groups: yup.array().of(yupTypes.namedNode()),
  source: yup.mixed().test({
    name: 'banner-source',
    message: 'web:banners.invalidSource',
    test(v: any) {
      const schema =
        typeof v === 'string' ? WebViewSourceSchema : ImageSourceSchema;
      try {
        schema.validateSync(v, { context: this.parent });
        return true;
      } catch (e) {
        return this.createError({
          message: (e as yup.ValidationError).message,
        });
      }
    },
  }),
});
