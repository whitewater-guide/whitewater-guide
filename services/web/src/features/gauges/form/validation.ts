import * as yup from 'yup';

import { CoordinateSchema, GaugeInputSchema } from '@whitewater-guide/commons';

import { GaugeFormData } from './types';
import { yupTypes } from '@whitewater-guide/validation';

const LoosePointSchema = yup
  .object({
    id: yup
      .string() // TODO: should be uuid, but i'm too lazy to make uuid without defined
      .nullable()
      .notRequired(),
    name: yup
      .string()
      .nullable()
      .notRequired(),
    description: yup
      .string()
      .nullable()
      .notRequired(),
    coordinates: CoordinateSchema.clone(),
    kind: yup
      .mixed()
      .oneOf(['gauge'])
      .notRequired(),
  })
  .strict(true)
  .noUnknown()
  .nullable();

export const GaugeFormSchema: yup.Schema<GaugeFormData> = GaugeInputSchema.clone().shape(
  {
    location: LoosePointSchema.clone(),
    requestParams: yupTypes.jsonString().nullable(),
  },
);
