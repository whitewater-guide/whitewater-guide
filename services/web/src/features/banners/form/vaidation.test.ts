import { LocalPhotoStatus } from '@whitewater-guide/clients';
import {
  BannerPlacement,
  BannerResolutions,
  createSafeValidator,
} from '@whitewater-guide/commons';
import { BannerFormData } from './types';
import { BannerFormSchema } from './validation';

type TestValue = [string, BannerFormData];
const validator = createSafeValidator(BannerFormSchema);

const correct: BannerFormData = {
  id: 'dddf9d28-c251-11e8-a355-529269fb1459',
  slug: 'test_banner',
  name: 'Some banner',
  priority: 10,
  enabled: true,
  placement: BannerPlacement.MOBILE_REGION_DESCRIPTION,
  source: 'https://banner.com',
  link: 'https://yamdex.ru',
  extras: '{ "foo": "bar" }',
  regions: [{ id: 'd10c02b6-c659-11e8-a355-529269fb1459', name: 'region' }],
  groups: [{ id: 'db178622-c659-11e8-a355-529269fb1459', name: 'group' }],
};

const correctValues: TestValue[] = [
  ['webview value', correct],
  [
    'local photo value',
    {
      ...correct,
      source: {
        id: 'foo',
        status: LocalPhotoStatus.READY,
        url: 'https://banner.com',
        resolution: BannerResolutions.get(
          BannerPlacement.MOBILE_REGION_DESCRIPTION,
        )!,
      },
    },
  ],
  ['null extras', { ...correct, extras: null }],
];

const incorrectValues: TestValue[] = [
  [
    'invalid resolution',
    {
      ...correct,
      source: {
        id: 'foo',
        status: LocalPhotoStatus.READY,
        url: 'https://banner.com',
        resolution: [100, 100],
      },
    },
  ],
];

it.each(correctValues)('should be valid for %s', (_, value) => {
  expect(validator(value)).toBeNull();
});

it.each(incorrectValues)('should be invalid for %s', (_, value) => {
  const result = validator(value);
  expect(result).not.toBeNull();
  expect(result).toMatchSnapshot();
});
