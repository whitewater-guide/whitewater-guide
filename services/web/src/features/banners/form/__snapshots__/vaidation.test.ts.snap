// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should be invalid for invalid resolution 1`] = `
Object {
  "source": Object {
    "key": "yup:photo.resolution",
    "options": Object {
      "height": 683,
      "width": 2048,
    },
  },
}
`;
